gr-uhd/grc/gen_uhd_usrp_blocks.py

```diff
@@ -154,65 +154,78 @@
-        self.${'$'}{id}.set_subdev_spec(${'$'}{${'sd_spec' + str(m)}}, ${m})
+        ########################################################################
+        ${'%'} if context.get('clock_source${m}')():
+        self.${'$'}{id}.set_clock_source(${'$'}{${'clock_source' + str(m)}}, ${m})
         ${'%'} endif
+        ########################################################################
         ${'%'} if context.get('time_source${m}')():
         self.${'$'}{id}.set_time_source(${'$'}{${'time_source' + str(m)}}, ${m})
         ${'%'} endif
-        ${'%'} if context.get('clock_source${m}')():
-        self.${'$'}{id}.set_clock_source(${'$'}{${'clock_source' + str(m)}}, ${m})
+        ########################################################################
+        ${'%'} if context.get('sd_spec${m}')():
+        self.${'$'}{id}.set_subdev_spec(${'$'}{${'sd_spec' + str(m)}}, ${m})
         ${'%'} endif
+        ########################################################################
         ${'%'} endif
-        % endfor
+        % endfor  # for m in range(max_mboards)
+        
+        self.${'$'}{id}.set_samp_rate(${'$'}{samp_rate})
+        
+        ${'%'} if sync == 'sync':
+        self.${'$'}{id}.set_time_unknown_pps(uhd.time_spec())
+        ${'%'} elif sync == 'pc_clock':
+        self.${'$'}{id}.set_time_now(uhd.time_spec(time.time()), uhd.ALL_MBOARDS)
+        ${'%'} else:
+        # No synchronization enforced.
+        ${'%'} endif
+        
         % for n in range(max_nchan):
         ${'%'} if context.get('nchan')() > ${n}:
         self.${'$'}{id}.set_center_freq(${'$'}{${'center_freq' + str(n)}}, ${n})
-        % if sourk == 'source':
-        ${'%'} if context.get('rx_agc${n}')() == 'Enabled':
-        self.${'$'}{id}.set_rx_agc(True, ${n})
-        ${'%'} elif context.get('rx_agc${n}')() == 'Disabled':
-        self.${'$'}{id}.set_rx_agc(False, ${n})
-        ${'%'} endif
         ${'%'} if context.get('rx_agc${n}')() != 'Enabled':
         ${'%'} if bool(eval(context.get('norm_gain' + '${n}')())):
         self.${'$'}{id}.set_normalized_gain(${'$'}{${'gain' + str(n)}}, ${n})
         ${'%'} else:
         self.${'$'}{id}.set_gain(${'$'}{${'gain' + str(n)}}, ${n})
         ${'%'} endif
+        ${'%'} endif  # if rx_agc${n} != 'Enabled'
+        
+        ${'%'} if context.get('ant${n}')():
+        self.${'$'}{id}.set_antenna(${'$'}{${'ant' + str(n)}}, ${n})
+        ${'%'} endif
+        ${'%'} if context.get('bw${n}')():
+        self.${'$'}{id}.set_bandwidth(${'$'}{${'bw' + str(n)}}, ${n})
+        ${'%'} endif
+        
+        % if sourk == 'source':
+        ${'%'} if context.get('rx_agc${n}')() == 'Enabled':
+        self.${'$'}{id}.set_rx_agc(True, ${n})
+        ${'%'} elif context.get('rx_agc${n}')() == 'Disabled':
+        self.${'$'}{id}.set_rx_agc(False, ${n})
         ${'%'} endif
+        self.${'$'}{id}.set_auto_dc_offset(${'$'}{${'dc_offs_enb' + str(n)}}, ${n})
```

Comment:
This call shouldn't always be called. There are USRPs which don't have this feature, and will then throw. Same with the next line.

Reply:
I fixed this in https://github.com/gnuradio/gnuradio/pull/3021.